File: Tarea5/krochanska.py
"""
D-Problem D
Nicolás Zapata Clavijo
8984273
"""

"""
En esta solución hago un BFS el cual tiene complejidad de O(n + m), sin embargo, lo ejecuto para cada estación importante por esto la complejidad de mi solución termina
siendo O(s(n + m)), siendo s el número de estaciones de importantes, n el número de nodos en el grafo y m el número de aristas.
"""
from sys import stdin
import heapq

INF = 10 ** 18

def dijkstra(adj, start, k):
    dist = [INF for _ in range(k)]
    dist[start] = 0
    heap = [(0, start)]
    while heap:
        d, u = heapq.heappop(heap)
        if d <= dist[u]:
            for v, w in adj[u]:
                if d + w < dist[v]:
                    dist[v] = d + w
                    heapq.heappush(heap, (dist[v], v))
    return dist

def main():
    t = int(stdin.readline().strip())
    for _ in range(t):
        n, s = map(int, stdin.readline().split())
        lineas = []
        apariciones = [0 for _ in range(n + 1)]
        for _ in range(s):
            linea = list(map(int, stdin.readline().split()))
            linea.pop()
            lineas.append(linea)
            for x in linea:
                apariciones[x] += 1
        importantes = []
        for i in range(1, n + 1):
            if apariciones[i] > 1:
                importantes.append(i)
        k = len(importantes)
        indice = {}
        for i in range(k):
            indice[importantes[i]] = i
        adj = [[] for _ in range(k)]
        for linea in lineas:
            ultimaImportante = -1
            distanciaAcumulada = 0
            for i in range(len(linea)):
                estacion = linea[i]
                if apariciones[estacion] > 1:
                    if ultimaImportante != -1:
                        u = indice[ultimaImportante]
                        v = indice[estacion]
                        costo = distanciaAcumulada * 2
                        adj[u].append((v, costo))
                        adj[v].append((u, costo))
                    ultimaImportante = estacion
                    distanciaAcumulada = 0  
                if i + 1 < len(linea):
                    distanciaAcumulada += 1
        mejorEstacion = importantes[0]
        mejorSuma = INF
        for i in range(k):
            dist = dijkstra(adj, i, k)
            suma = 0
            for j in range(k):
                suma += dist[j]
            if suma < mejorSuma or (suma == mejorSuma and importantes[i] < mejorEstacion):
                mejorSuma = suma
                mejorEstacion = importantes[i]
        print("Krochanska is in: " + str(mejorEstacion))

main()

File: Tarea5/krochanska.cpp
#include <bits/stdc++.h>
using namespace std;

const long long INF = 1e18;

vector<long long> dijkstra(vector<vector<pair<int,int>>> &adj, int start, int k){
    vector<long long> dist(k, INF);
    priority_queue<pair<long long,int>, vector<pair<long long,int>>, greater<pair<long long,int>>> pq;
    dist[start] = 0;
    pq.push({0, start});
    while(!pq.empty()){
        long long d = pq.top().first;
        int u = pq.top().second;
        pq.pop();
        if(d <= dist[u]){
            for(auto &edge : adj[u]){
                int v = edge.first;
                int w = edge.second;
                if(d + w < dist[v]){
                    dist[v] = d + w;
                    pq.push({dist[v], v});
                }
            }
        }
    }
    return dist;
}

int main(){
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int t; cin >> t;
    while(t--){
        int n, s; cin >> n >> s;
        vector<vector<int>> lineas(s);
        vector<int> apariciones(n+1, 0);
        for(int i = 0; i < s; i++){
            int x;
            while(cin >> x && x != 0){
                lineas[i].push_back(x);
                apariciones[x]++;
            }
        }
        vector<int> importantes;
        for(int i = 1; i <= n; i++){
            if(apariciones[i] > 1){
                importantes.push_back(i);
            }
        }
        int k = importantes.size();
        unordered_map<int,int> indice;
        for(int i = 0; i < k; i++){
            indice[importantes[i]] = i;
        }
        vector<vector<pair<int,int>>> adj(k);
        for(auto &linea : lineas){
            int ultima = -1, distAcum = 0;
            for(int i = 0; i < linea.size(); i++){
                int est = linea[i];
                if(apariciones[est] > 1){
                    if(ultima != -1){
                        int u = indice[ultima];
                        int v = indice[est];
                        int costo = distAcum * 2;
                        adj[u].push_back({v, costo});
                        adj[v].push_back({u, costo});
                    }
                    ultima = est;
                    distAcum = 0;
                }
                if(i + 1 < linea.size()){
                    distAcum++;
                }
            }
        }
        int mejorEstacion = importantes[0];
        long long mejorSuma = INF;
        for(int i = 0; i < k; i++){
            vector<long long> dist = dijkstra(adj, i, k);
            
            long long suma = 0;
            for(int j = 0; j < k; j++){
                suma += dist[j];
                if(suma >= mejorSuma){
                    j = k;
                }
            }
            if(suma < mejorSuma || (suma == mejorSuma && importantes[i] < mejorEstacion)){
                mejorSuma = suma;
                mejorEstacion = importantes[i];
            }
        }
        cout << "Krochanska is in: " << mejorEstacion << "\n";
    }
}